### PR TITLE
implement bidirectional lstm

### PR DESCRIPTION
**Pull request by wcshds**
I need bidirectional lstm in CRNN model.

### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### CODE REVIEW

File: burn-core/src/nn/rnn/lstm.rs
Comment (nathanielsimard) [inline burn-core/src/nn/rnn/lstm.rs:63]: Even if it's a boolean condition, I feel like a match might be easier to read, but using `new_gate` is a big win.

File: burn-core/src/nn/rnn/lstm.rs
Comment (nathanielsimard) [inline burn-core/src/nn/rnn/lstm.rs:127]: I feel like it would be better to have a `BiDirectionalLstm` module instead of one module that does both. The bi-directional module might just have two Lstm modules instead of duplicating all the gates. We could also make it more Rusty by using Kind types:

```rust
pub trait LstmKind<B: Backend> {
    fn forward(...) -> ...;
}
#[derive(Module)]
pub struct UniDirectional{
    input_gate: GateController<B>,
    forget_gate: GateController<B>,
    output_gate: GateController<B>,
    cell_gate: GateController<B>,
};

#[derive(Module)]
pub struct BiDirectional<B: Backend> {
   forward: Lstm<B, UniDirectional>,
   backward: Lstm<B, UniFirectional>,
}
pub struct Lstm<B: Backend, K: LstmKind<B> = UniDirectional<B>> {
   state: K,
}

impl<B: Backend, K: LstmKind<B>> Lstm<B, K> {
   pub fn forward(...) -> ... {
       self.kind.forward(...) // static dispatch to the right forward pass depending on the kind.
   }
}

impl<B: Backend> LstmKind<B> for UniDirectional<B> {
   fn forward(...) -> ... {
       // uni directional forward pass
   }
}
impl<B: Backend> LstmKind<B> for BiDirectional<B> {
   fn forward(...) -> ... {
       // bi directional forward pass
   }
}
```

This is equivalent to having two different modules for the uni directional and bi directional lstm but with syntax sugar so users can also use `Lstm<B>` for uni directional or `Lstm<B, BiDirectional>` when they want both directions. This is the same pattern that we use for the Tensor API.

Comment (nathanielsimard) [reply burn-core/src/nn/rnn/lstm.rs:127]: This would also make it backward compatible in term of code, but not in term of state (Recorder).

Comment (nathanielsimard) [reply burn-core/src/nn/rnn/lstm.rs:127]: @agelas what are your thoughts

Comment (agelas) [reply burn-core/src/nn/rnn/lstm.rs:127]: @nathanielsimard I agree that this pattern fits better with the overall approach that is usually taken. I think composing LSTM modules like this is a bit more elegant than having various toggles in the `forward` pass, and we can reuse most of the prior implementation that way.

File: burn-core/src/nn/rnn/lstm.rs
Comment (agelas) [inline burn-core/src/nn/rnn/lstm.rs:154]: @nathanielsimard Is this necessary? The implementation is already tied to a device vis-à-vis the backend right?

Comment (wcshds) [reply burn-core/src/nn/rnn/lstm.rs:154]: In my understanding, `Module::fork` can only fork the tensor in the module struct field to the given device. So, when creating a new Tensor in the `forward()`, it is necessary to specify the device?

By the way, now I feel that bidirectional LSTM may not be a common requirement for everyone. If someone needs bidirectional LSTM or multi-layer LSTM, it's probably best for them to implement it themselves. Additionally, keeping the implementation of Lstm unchanged is essential to avoid disrupting code compatibility.

File: burn-core/src/nn/rnn/lstm.rs
Comment (wcshds) [reply burn-core/src/nn/rnn/lstm.rs:127]: @nathanielsimard The problem I have is that the `forward()` for `UniDirectional` and `BiDirectional` have different function signatures due to the inconsistency in the shape of the `state`. This seems not easy to resolve, so it's probably best to keep the LSTM implementation unchanged and let users implement bidirectional LSTM or multi-layer LSTM on their own.

Comment (louisfd) [reply burn-core/src/nn/rnn/lstm.rs:127]: Not sure how often bidirectional LSTM is used, but could we have an implementation totally separated from unidirectional LSTM, and just offer both? Because I do agree that messing with function signatures might get a bit bothersome

Comment (nathanielsimard) [reply burn-core/src/nn/rnn/lstm.rs:127]: The pattern I'm proposing creates two different types, so there are no problems with the type signature. It's just a way for both types to use the LSTM nomenclature. It's probably easier to start with two different types, then we can add the pattern afterward if it makes sense.

File: burn-core/src/nn/rnn/lstm.rs
Comment (nathanielsimard) [reply burn-core/src/nn/rnn/lstm.rs:154]: The device is necessary when creating a new tensor; otherwise, the new tensor will be created on the default device, but not necessarily on the same device as the module.

File: burn-core/src/nn/rnn/lstm.rs
Comment (wcshds) [reply burn-core/src/nn/rnn/lstm.rs]: @nathanielsimard This pattern is a bit tricky for me, so I think it's best for me to create a new type for bidirectional LSTM. I sincerely hope you can improve it if possible. Thank you very much!

File: burn-core/src/nn/rnn/lstm.rs
Comment (nathanielsimard) [inline burn-core/src/nn/rnn/lstm.rs:219]: I guess it wasn't possible to use two Lstm modules here?

**Comment by wcshds** [comment]
I think the current implementation of bidirectional lstm can work during inference, but the implementation cannot update the parameters of the model in backward propagation due to #1098. Using `Tensor::cat` instead of `Tensor::slice_assign` can solve this problem, but it seems to have some negative performance impact, so I think it's best to wait for the `Tensor::slice_assign` bug to be resolved for now.

**Comment by antimora** [comment]
Closing this ticket and linking to this ticket: https://github.com/tracel-ai/burn/issues/1537. So someone else can pick up.

**Comment by wcshds** [comment]
@antimora Could you reopen this pull request? The issue regarding Autodiff has been resolved in #1575, and I think it's time to proceed with implementing the Bidirectional LSTM.

**Comment by wcshds** [comment]
Wgpu test failed, but I don't know why...

**Comment by agelas** [comment]
What was the error in wgpu?

**Comment by agelas** [comment]
Hm ok well on ubuntu-22.04 everything seems fine when I run the tests locally. Given that ubuntu and windows work, maybe it hints at possible compatibility issues or differences in how Metal handles resource management compared to Vulkan (ie Linux) or DirectX (Windows)?

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (nathanielsimard) [inline crates/burn-core/src/nn/rnn/lstm.rs:82]: Eventhough that's a breaking change, I think it might be beneficial to create a type for the state.

```rust
pub struct LstmState {
    pub cell: Tensor<B, 2>,
    pub hidden: Tensor<B, 2>,
}
```

We can remove the optional and implement `Default` instead. It also gives us a space to document what each element of the state is used for.

We could also do the same for the return type, where it's easy to make mistake (chosing the wrong returned tensor).

What do you think?

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (nathanielsimard) [inline crates/burn-core/src/nn/rnn/lstm.rs:106]: Not linked to the current change, but is zero necessary or empty would be enough here?

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (nathanielsimard) [inline crates/burn-core/src/nn/rnn/lstm.rs:451]: Just a detail, but we can use `&burn_tensor::Device<TestBackend>` instead of this notation, just a bit prettier, especially when importing `burn_tensor::Device`.

**Review by nathanielsimard — COMMENTED** [review]
I like it! I have a few comments, but nothing major. Would also like @laggui to review this.

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (laggui) [reply crates/burn-core/src/nn/rnn/lstm.rs:82]: I quite like that idea, especially for the return type!

**Review by laggui — COMMENTED** [review]
Great rewrite to add the `BiLstm`! I don't see any issues 🙂  Just a minor comment regarding the state as pointed out by Nath.

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (wcshds) [reply crates/burn-core/src/nn/rnn/lstm.rs:82]: If breaking changes are acceptable, I'd like to follow PyTorch's return values `output, (h_n, c_n)`.

image

In that case, we can return `(output, state)`. For `LSTM`, `output` and `state.hidden` are the same; for `BiLSTM`, `output` is a concatenated tensor, and the cell state and hidden state can be provided in `state` to the user without modification.

Comment (wcshds) [reply crates/burn-core/src/nn/rnn/lstm.rs:82]: If you agree to the modifications according to the PyTorch outputs, `batched_cell_state` is unnecessary. For the state, we only need to return the hidden state and cell state of the last time step.

**Comment by wcshds** [comment]
The outputs of both `Lstm` and `BiLstm` are now aligned with PyTorch.

**Comment by wcshds** [comment]
I changed the batch size to 2, and the test passed. This seems to be a bug with the Wgpu backend, so I've opened an issue #1656.

**Review by nathanielsimard — APPROVED** [review]
No more comments from my side, great job! We will investigate the wgpu test problem, but it's quite unreleated.

**Comment by antimora** [comment]
We will merge it, once we have an approval from @laggui

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (laggui) [inline crates/burn-core/src/nn/rnn/lstm.rs:82]: Since we document the return types, I think we can simplify the docstring here to:

Applies the forward pass on the input tensor. This LSTM implementation returns the state for each element in a sequence (i.e., across `seq_length`) and a final state.

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (laggui) [inline crates/burn-core/src/nn/rnn/lstm.rs:217]: Same as with the `Lstm` forward docstring. We can simplify to:

Applies the forward pass on the input tensor. This Bidirectional LSTM implementation returns the state for each element in a sequence (i.e., across `seq_length`) and a final state.

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (laggui) [inline crates/burn-core/src/nn/rnn/lstm.rs:410]: Dead comment, we can remove 🙂

**Review by laggui — CHANGES_REQUESTED** [review]
Good job! 😄  I like the addition of `LstmState` and the returned output.

Only a few minor changes, and then we can merge 🎉

File: crates/burn-core/src/nn/rnn/lstm.rs
Comment (wcshds) [reply crates/burn-core/src/nn/rnn/lstm.rs]: All done.

**Pull request merged**
